src/quicklist.c

```diff
@@ -718,6 +718,16 @@ void quicklistDelEntry(quicklistIter *iter, quicklistEntry *entry) {
             iter->current = prev;
             iter->offset = -1;
         }
+
+        /* Also updates param 'entry' to the right entry of the deleted one. */
+        entry->node = next;
```

Comment:
Doesn't this need to depend on the direction like the code above? 
Same for `lpFirst` below.

Reply:
Because index is independent of direction, i realize that I was going about it all wrong.
After second thought, I think it would be futile to make effort to keep `key->u.list.index` in sync with `key->u.list.entry`.
For example a list (a, b, c, d, e)
If index is 2(c), delete it and entry needs to point to d.
If index is -3(c), deleting it and entry needs to point to b.
But inside listTypeDelete there is no way to know what the index is.

Reply:
@oranagra @zuiderkwast It seems i started off on the wrong foot.
We should sync entry and index to iter, instead of thinking about how to synchronize entry and index.
In [`86c61ac` (#11383)](https://github.com/redis/redis/pull/11383/commits/86c61ac49f0890b86333ef101f1761087828a747), I use listTypeNext to synchronize entry to iter's current entry, and at the same time update the index.

Reply:
> * Change other things in t_list.c or quicklist.c iterator logic and API to make them more simple and support the operations we need (delete, insert, replace without invalidating the iterator). It's fine that we require something like getCurrent() after an element has changed, so we don't always need to keep the current entry up to date. (This is probably a larger change.)

This is actually something I've always wanted to do, but it's too complicated because the insert operation involves node merging and splitting, and it takes a lot of work to keep the iterator in state.

Reply:
@sundb please re-do the test suite changes your force-push wiped..
specifically the reverse iteration.
maybe even extend them to cover the case of deleting the entire list, and maybe somehow cover quicklist with more than one node.

Reply:
Done

Reply:
LGTM.
@zuiderkwast please take a look.

the only additional thought i have is that maybe the tests can somehow be more aggressive, and validate the index better after deletion.
i.e. the `list.getall` operation we do doesn't validate the index of the iterator.
maybe we can change `list.edit` to return the index, or add another "command" argument that returns an element in the reply (i.e. like `k` but instead of skip it returns the element).
then all the calls to `list.edit` will be changed to validate the next item in the iteration.

another random idea is to add a fuzzer, but then we need to match it against another implementation (same as `ZDIFF fuzzing` test matches the result of two different implementations).
i guess that's too much effort.